pkg/server/service.go

```diff
@@ -123,14 +135,79 @@ func NewCRIContainerdService(containerdEndpoint, rootDir, networkPluginBinDir, n
 		return nil, fmt.Errorf("failed to create stream server: %v", err)
 	}
 
+	c.eventMonitor, err = newEventMonitor(c)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create event monitor: %v", err)
+	}
+
+	// Create the grpc server and register runtime and image services.
+	c.server = grpc.NewServer()
+	runtime.RegisterRuntimeServiceServer(c.server, newInstrumentedService(c))
+	runtime.RegisterImageServiceServer(c.server, newInstrumentedService(c))
+
 	return newInstrumentedService(c), nil
 }
 
-func (c *criContainerdService) Start() {
-	c.startEventMonitor()
+// Run starts the cri-containerd service.
+func (c *criContainerdService) Run() error {
+	glog.V(2).Info("Start cri-containerd service")
+	// TODO(random-liu): Recover state.
+
+	// Start event handler.
+	glog.V(2).Info("Start event monitor")
+	eventMonitorClose := c.eventMonitor.start()
+
+	// Start streaming server.
+	glog.V(2).Info("Start streaming server")
+	streamServerClose := make(chan struct{})
 	go func() {
 		if err := c.streamServer.Start(true); err != nil {
 			glog.Errorf("Failed to start streaming server: %v", err)
 		}
+		close(streamServerClose)
+	}()
+
+	// Start grpc server.
+	// Unlink to cleanup the previous socket file.
+	glog.V(2).Info("Start grpc server")
+	err := syscall.Unlink(c.serverAddress)
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("failed to unlink socket file %q: %v", c.serverAddress, err)
+	}
+	l, err := net.Listen(unixProtocol, c.serverAddress)
+	if err != nil {
+		return fmt.Errorf("failed to listen on %q: %v", c.serverAddress, err)
+	}
+	grpcServerClose := make(chan struct{})
+	go func() {
+		if err := c.server.Serve(l); err != nil {
+			glog.Errorf("Failed to serve grpc grpc request: %v", err)
+		}
+		close(grpcServerClose)
```

Comment:
if we close - won't the criContainerdService stop below?

Reply:
Yeah, that's what we want.

We want criContainerdService to exit when a critical service is down.

Reply:
@Random-Liu one question. Is there any code path that would have closed the channel already ? Also is there a code path where we would end up being blocked on c.server.Serve ? Just wondering if this go routine could get stuck and also ensuring we dont double close the channel.

Reply:
We only close the channel here, we don't even pass it to outside of the function. :)